sdk/src/main/java/com/hedera/hashgraph/sdk/TopicMessageQuery.java

```diff
@@ -1,110 +1,199 @@
         return this;
     }
 
     public TopicMessageQuery setEndTime(Instant endTime) {
+        Objects.requireNonNull(endTime, "endTime must not be null");
         builder.setConsensusEndTime(InstantConverter.toProtobuf(endTime));
         return this;
     }
 
     public TopicMessageQuery setLimit(long limit) {
+        if (limit < 0) {
+            throw new IllegalArgumentException("limit must be positive");
+        }
         builder.setLimit(limit);
         return this;
     }
 
+    public TopicMessageQuery setCompletionHandler(Runnable completionHandler) {
+        Objects.requireNonNull(completionHandler, "completionHandler must not be null");
+        this.completionHandler = completionHandler;
+        return this;
+    }
+
     public TopicMessageQuery setErrorHandler(BiConsumer<Throwable, TopicMessage> errorHandler) {
+        Objects.requireNonNull(errorHandler, "errorHandler must not be null");
         this.errorHandler = errorHandler;
         return this;
     }
 
-    // TODO: Refactor into a base class when we add more mirror query types
-    public SubscriptionHandle subscribe(
-        Client client,
-        Consumer<TopicMessage> onNext
-    ) {
-        SubscriptionHandle subscriptionHandle = new SubscriptionHandle();
+    public TopicMessageQuery setMaxAttempts(int maxAttempts) {
+        if (maxAttempts < 0) {
+            throw new IllegalArgumentException("maxAttempts must be positive");
+        }
+        this.maxAttempts = maxAttempts;
+        return this;
+    }
 
-        makeStreamingCall(client, subscriptionHandle, errorHandler, builder.build(), onNext, 0, new Instant[]{null});
+    public TopicMessageQuery setMaxBackoff(Duration maxBackoff) {
+        if (maxBackoff == null || maxBackoff.toMillis() < 500L) {
+            throw new IllegalArgumentException("maxBackoff must be at least 500 ms");
+        }
+        this.maxBackoff = maxBackoff;
+        return this;
+    }
 
-        return subscriptionHandle;
+    public TopicMessageQuery setRetryHandler(Predicate<Throwable> retryHandler) {
+        Objects.requireNonNull(retryHandler, "retryHandler must not be null");
+        this.retryHandler = retryHandler;
+        return this;
     }
 
-    private static void makeStreamingCall(
-        Client client,
-        SubscriptionHandle subscriptionHandle,
-        @Nullable BiConsumer<Throwable, TopicMessage> errorHandler,
-        ConsensusTopicQuery query,
-        Consumer<TopicMessage> onNext,
-        int attempt,
-        // startTime must be `final` or `effectively final` to be used within closures.
-        Instant[] startTime
-    ) {
-        if (attempt > 10) {
-            if (errorHandler != null) {
-                errorHandler.accept(new Error("Failed to connect to mirror node"), null);
-                return;
-            }
+    private void onComplete() {
+        TopicId topicId = TopicId.fromProtobuf(builder.getTopicID());
```

Comment:
Can we use `var` instead of consistency's sake; I'm just going to make one comment here, but there are other places where this is also the case.

Reply:
Sure, switched most variables I added to `var`.